### ISSUE TITLE
[Bug]: Next.js module transpilation is not supported

### ISSUE THREAD

**Issue by rodrigo-arias** [storybookjs/storybook#22070]
### Describe the bug

Despite the [Next.js + Storybook](https://storybook.js.org/addons/storybook-addon-next) addon description claims _"What Next.js supports out of the box, this addon makes possible in Storybook"_ the [Next.js Built-in module transpilation](https://nextjs.org/blog/next-13-1#built-in-module-transpilation-stable) doesn't seem to be working.

I have some un-transpiled ES6 packages that previously used [next-transpile-modules](https://www.npmjs.com/package/next-transpile-modules) and now use the built-in the Next.js configuration. These work fine in Next.js but fail in Storybook after migrating to v7.

Previously I was using the following Webpack custom configuration

```
webpackFinal: async (config) => {
  config.module.rules[0].exclude = [/node_modules\/(?!@my-scope).*/]
  return config
}
```

The Next.js addon doesn't seem to detect the `transpilePackages` attribute from the `next.config.js` nor does the custom Webpack rule work.

### To Reproduce

1. Install a dependency with un-transpiled ES6 code.
2. Add this dependency to be transpiled into Next.js as per https://nextjs.org/blog/next-13-1#built-in-module-transpilation-stable
3. Verify that Next.js runs correctly.
4. Verify that Storybook fails by not transpiling the dependency. You will see an error similar to _"You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file."_

### System

```shell
System:
  OS: macOS 13.2
  CPU: (8) arm64 Apple M1 Pro
Binaries:
  Node: 16.18.1 - /opt/homebrew/opt/node@16/bin/node
  npm: 8.19.2 - /opt/homebrew/opt/node@16/bin/npm
Browsers:
  Chrome: 112.0.5615.49
  Safari: 16.3
npmPackages:
  @storybook/addon-a11y: ^7.0.4 => 7.0.4 
  @storybook/addon-essentials: ^7.0.4 => 7.0.4 
  @storybook/addon-links: ^7.0.4 => 7.0.4 
  @storybook/addon-mdx-gfm: ^7.0.4 => 7.0.4 
  @storybook/manager-api: ^7.0.4 => 7.0.4 
  @storybook/nextjs: ^7.0.4 => 7.0.4 
  @storybook/react: ^7.0.4 => 7.0.4 
  @storybook/theming: ^7.0.4 => 7.0.4
```

**Comment by shilman** [comment]
Does the previously-used `webpackFinal` workaround solve the problem in v7?

**Comment by rodrigo-arias** [comment]
@shilman no, neither the auto-detected configuration nor the Webpack workaround seems to work.

**Comment by linkb15** [comment]
I am facing this issue with css exported in internal bundling is not being transpiled when running storybook in nextjs, hence the styling is not applying correctly.

was running storybook with react-vite as well but it is applying since vite did the transpiling automatically. however the webpack5 that storybook/nextjs have still need to be configured manually. 

Now I am struggling to configure the webpack5 configuration to handle the internal transpiling. Need help if someone managed to solve this earlier.

**Comment by awinograd** [comment]
I'm running into the same issue. I did try shoehorning `next-transpile-modules` into the storybook/nextjs webpack compilation workflow and got it to print that it was transpiling the expected modules, but surprisingly still ended up with errors that some of my files were not actually transpiled. pseudo code was something like:

```ts

const withTM = require('next-transpile-modules')(transpileModules, {
  resolveSymlinks: true,
  debug: true,
})
const tmWebpackCallback = withTM({}).webpack

...

  webpackFinal: async (config) => {
     return tmWebpackCallback(cfg, { defaultLoaders: [] })
  }

```

**Comment by dmalament-wheelsup** [comment]
> I am facing this issue with css exported in internal bundling is not being transpiled when running storybook in nextjs, hence the styling is not applying correctly.
> 
> was running storybook with react-vite as well but it is applying since vite did the transpiling automatically. however the webpack5 that storybook/nextjs have still need to be configured manually.
> 
> Now I am struggling to configure the webpack5 configuration to handle the internal transpiling. Need help if someone managed to solve this earlier.

@linkb15 I'm having this exact same issue. Were you able to figure it out?

**Comment by felipeegasneris** [comment]
same issue here, any solutions??? must we configure webpackfinal manually to transpile?

**Comment by salazarm** [comment]
same

**Comment by dmalament-wheelsup** [comment]
Yes I was able to solve this thanks to https://github.com/nrwl/nx/issues/16962#issuecomment-1564747886

My code in `webpack.config.js` for storybook: 
```typescript
    // this section is necessary to turn off tree shaking for css files. without this, none of the CSS from wu-react-components was being loaded in storybook. https://webpack.js.org/guides/tree-shaking/#mark-the-file-as-side-effect-free
    // credit to https://github.com/nrwl/nx/issues/16962#issuecomment-1564747886 for this solution

    const { module } = config;

    const cssRule = module.rules.find((rule) => rule.test.toString().includes('css'));
    cssRule.sideEffects = true
```

**Comment by salazarm** [comment]
I tried using:
```
framework: {
    name: '@storybook/nextjs',
    options: {
      transpilePackages: ['@dagster-io/ui-components', '@dagster-io/ui-core/'],
    },
  },
```

But I end up with:

```
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
| import {IconName, Icon} from './Icon';
| 
> export type AlertIntent = 'info' | 'warning' | 'error' | 'success';
| 
| interface Props {
```

**Comment by salazarm** [comment]
I added:
```
{
            test: /\.(ts|tsx?|js|jsx?)$/,
            use: {
              loader: 'babel-loader',
              options: {
                presets: ['next/babel'],
              },
            },
          },
```

To my rules and that seemed to work for the most part.

**Comment by dstaley** [comment]
> I added:
> 
> ```
> {
>             test: /\.(ts|tsx?|js|jsx?)$/,
>             use: {
>               loader: 'babel-loader',
>               options: {
>                 presets: ['next/babel'],
>               },
>             },
>           },
> ```
> 
> To my rules and that seemed to work for the most part.

Unfortunately this breaks automatic `ArgsTypes` generation for TypeScript files. I've had better results with the following, which adjusts the default `exclude` condition to _include_ specific packages (in this case all packages within the `@dagster-io` scope):

```tsx
webpackFinal: async (config) => {
	const babelLoader = config.module.rules.find((r) => {
		return (
			r.use &&
			r.use.length > 0 &&
			r.use[0].loader &&
			r.use[0].loader.includes('babel-loader')
		)
	})
	babelLoader.exclude = [
		/node_modules\/(?!@dagster-io)/,
		...babelLoader.exclude.filter((e) => e.toString() !== '/node_modules/'),
	]

	return config
}
```

**Comment by filipjakov** [comment]
Anyone managed to find a solution? Encountered the problem and have no idea on how to fix it 🤷

**Comment by rodrigo-arias** [comment]
@filipjakov I didn't find a solution and doubt Storybook will fix it. I finally transpile the packages because I controlled them, but I know that's not always an option.

**Comment by valentinpalkovic** [comment]
I’ll try to find a solution in the upcoming weeks! Stay tuned!

**Comment by r34son** [comment]
> I’ll try to find a solution in the upcoming weeks! Stay tuned!

Any updates?

**Comment by valentinpalkovic** [comment]
Please try out our freshly baked experimental `@storybook/experimental-nextjs-vite` package. It is based on Vite and removes the clutter around Webpack and Babel.

For more information, please follow our documentation:
https://storybook.js.org/docs/get-started/frameworks/nextjs#with-vite

Very likely the `transpilePackages` attribute from the next.config is respected.

**Comment by r34son** [comment]
> Please try out our freshly baked experimental `@storybook/experimental-nextjs-vite` package. It is based on Vite and removes the clutter around Webpack and Babel.
> 
> For more information, please follow our documentation:
> https://storybook.js.org/docs/get-started/frameworks/nextjs#with-vite
> 
> Very likely the `transpilePackages` attribute from the next.config is respected.

What about nextjs optimizePackageImports feature? Does nextjs-vite support it? I have another issue about that https://github.com/storybookjs/storybook/issues/30126

**Comment by valentinpalkovic** [comment]
That's the thing I am not sure about. We are passing a lot of next.js config options to its SWC transpiler. It may work out of the box.

**Comment by valentinpalkovic** [comment]
@r34son Have you tried out `@storybook/experimental-nextjs-vite`?

**Comment by r34son** [comment]
> @r34son Have you tried out `@storybook/experimental-nextjs-vite`?

Not yet

**Comment by wvan1901** [comment]
I also ran into this issue and I have this workaround for the project I'm working on, this works for our project but your outcome may vary. I can't provide the Repo but I can provide the main.ts file:
```ts
import type {StorybookConfig} from '@storybook/nextjs'
const config: StorybookConfig = {
  // ... other configs
  webpackFinal: async (config) => {
    if (config.module && config.module.rules) {
      config.module.rules.push({
        test: /\.(ts|tsx)$/,
        loader: require.resolve('swc-loader'),
        options: {
          jsc: {
            transform: {
              react: {
                runtime: 'automatic'
              }
            }
          }
        }
      });
    }
    if (config.resolve && config.resolve.extensions) {
      config.resolve.extensions.push('.ts', '.tsx');
    }
    return config;
  }
};
export default config;
```

**Comment by valentinpalkovic** [comment]
Be cautious with this workaround though. Next.js uses a custom compiled and enhanced version of SWC. The swc-loader wouldn’t pick it up and therefore a lot of issues may appear with this solution.

**Comment by r34son** [comment]
> @r34son Have you tried out `@storybook/experimental-nextjs-vite`?

Tried, but unfortunately having some issues with resolving modules

**Comment by GriffinSauce** [comment]
> [..] 
> Unfortunately this breaks automatic `ArgsTypes` generation for TypeScript files. I've had better results with the following, which adjusts the default `exclude` condition to _include_ specific packages (in this case all packages within the `@dagster-io` scope):
> [..]

Thanks @dstaley 
A small modification of this did it for me, I had to adjust `includes` as well:
```.js
  const babelLoader = config.module.rules.find(rule =>
    rule?.use?.[0]?.loader?.includes('babel-loader')
  )
  babelLoader.include = [path.resolve('../../packages')]
  babelLoader.exclude = [
    /node_modules\/(?!@our-monorepo-scope)/,
    ...babelLoader.exclude.filter(e => e.toString() !== '/node_modules/'),
  ]
```

**Another tip for those with monorepo's**: this framework relies on aliasing the `next` dependency, so if you have any tooling that aliases package dependencies across your monorepo you'll want to exclude `next` from that.

---

About `@storybook/experimental-nextjs-vite` - this would mean migrating any other adjustments we made* + managing another bundler that we don't use for production. Is the idea that this would become the default? I'm guessing it'll work beautifully for default behaviours but could get painful when any customizations are involved.

*for example: using Linaria for styling

**Comment by valentinpalkovic** [comment]
I can't reproduce it. Neither with `lodash-es`, nor with `@dagster-io/ui`. `lodash-es` doesn't throw in Storybook and components from `@dagster-io/ui` just render fine. Can someone please provide a reliable reproduction (Repository, or exact reproduction steps)
